### DISCUSSION TITLE
Add Support for `seqcli create user type=Entra`

### DISCUSSION THREAD

**Discussion by Mike-E-angelo** [datalust/seq-tickets#2068] [Ideas]
Reference discussion:
https://github.com/datalust/seq-tickets/discussions/2067

For automation purposes, it would be super valuable to be able to create a new Entra ID user and assign it as the administrator of the new server instance.  Currently the only way this can be done is with user/pass which at this point in time seems a little ol'-fashioned. 😁

Thank you for any consideration. 🙏

**Comment by nblumhardt** [comment]
Hi Michael,

I think this will/should be possible with Seq 2024.1 and its corresponding `seqcli`.

The basic workflow will be:

- Set a username and password on first run, i.e. with `SEQ_FIRSTRUN_*`, as per the normal container deployment process
  * The username should be your AAD one, e.g. `example@microsoftonline.com` or similar
- Use those credentials with `seqcli` to create an API key with admin permissions
- Use the `seqcli setting set` command to configure values for the AAD provider (but don't enable it yet)
  * `AzureADAuthority` (I think this should be `https://login.microsoftonline.com` IIRC)
  * `AzureADClientId`
  * `AzureADClientKey`
  * `AzureADTenantId`
- Finally, set `AuthenticationProvider` to `"Azure Active Directory"

This should be enough to get things going, though I can't check through right now.

(Guess we'll need to create some `Entra*` aliases for all of these sometime soon 😅 )

HTH!

**Reply by Mike-E-angelo** [reply]
Thank you very much @nblumhardt for the guidance, it is greatly appreciated.  I have been able to get all the code in place, but there does seem to be a gotcha with `seqcli setting set` ... namely, it does not seem to exist. :D  Do you mean `seqcli config` ?  That's the closest I could find when [reviewing the commands](https://github.com/datalust/seqcli#commands).  I did try to use these commands but they do not appear to accept an API key, nor do they accept a server it would seem. Please let me know if I have something fundamentally misunderstood.

**Reply by liammclennan** [reply]
Hi @Mike-E-angelo 

Nick meant `seqcli config set` which should accept server and API key values. It doesn't make sense for a `seqcli` command to not have a server argument. I will investigate why these options are missing from the docs.

**Reply by Mike-E-angelo** [reply]
Hey there @liammclennan 👋 thank you for that suggestion.  I did try to poke around with your suggested call, but this too also appears to not exist:

![image](https://github.com/datalust/seq-tickets/assets/2931376/b4b99389-50f5-433b-8c19-ffa12bd68363)

Please let me know what I have misunderstood.

**Reply by liammclennan** [reply]
Quite right. Try like this:

seqcli config -k connection.serverUrl -v https://your-seq-server
seqcli config -k connection.apiKey -v your-api-key

**Reply by Mike-E-angelo** [reply]
OK small victories... those seem to be local settings and seem to work.  If I understand correctly, however, the settings above such as `AzureADAuthority` and `AzureADClientId` are meant to be applied to the server I would like to authenticate with via AAD/Entra.  If I attempt to use these values I get an exception:
![image](https://github.com/datalust/seq-tickets/assets/2931376/6509ced2-d150-4883-9ead-e50bf5f56937)

**Reply by nblumhardt** [reply]
Sorry folks, I should have underlined the 2024.1 aspect - unless you're running the unreleased preview of `seqcli` there's no `setting` command.

https://github.com/datalust/seq-tickets/discussions/2068#discussioncomment-8213750

(`config` is just for `seqcli`'s local config file.)

HTH!

**Reply by nblumhardt** [reply]
Sorry, I posted the wrong link; it's https://github.com/datalust/seqcli/pull/299

**Reply by Mike-E-angelo** [reply]
Ah!  Got it!  Mystery solved.  Yes, the 2024.1 is plainly there in your first sentence, but with my use of `datalust/seq:latest` I thought that was it. :D  OK no biggie I can get these changes shelved in the meantime and revisit once it is available.  Thank you again so much for taking the time to support an individual license. :) 👍🙏

**Reply by Mike-E-angelo** [reply]
Alright @nblumhardt with 2024.1 everything seems to be working now with setting values. 👍  I have created an application registration with a redirect of `https://SEQ_API_CANONICALURI/aad`.  Everything appears to be in order.  However, when I press `Log in using Azure Active Directory` on the front page, it redirects me here:

```
https://SEQ_API_CANONICALURI/oidc/challenge?toLocalUrl=
```

With the following error:

```
{"Error":"An unhandled error occurred while serving the request (token: <omitted>)."}
```

Not sure if token is protected so I have omitted it for now.  Please let me know of any further information I can provide to assist.  I will poke/prod around here in the meantime.

**Reply by Mike-E-angelo** [reply]
Please disregard the above!  I had a "durp" moment after posting it and went into the logs of the file share where the problem was, well, logged. 🤦‍♂️  It appears the `seqcli setting set -n=AzureADTenantId -v=$TenantId` resolved to `$TenantId` as the tenant.  Using `seqcli setting set -n=AzureADTenantId -v="$TenantId"` and seeing if that works better.  I think I have enough here to figure out how to proceed and will update here with success.  Hopefully. 😄

**Reply by Mike-E-angelo** [reply]
SUCCESS! :D  WOOHOO!  That is awesome.  Everything works!  This is what I ended up using:

```
seqcli setting set -n=AzureADAuthority -v="login.microsoftonline.com" -s $Address --apikey=$system
seqcli setting set -n=AzureADClientId -v="$application" -s $Address --apikey=$system
seqcli setting set -n=AzureADClientKey -v="$secret" -s $Address --apikey=$system
seqcli setting set -n=AzureADTenantId -v="$TenantId" -s $Address --apikey=$system
seqcli setting set -n=AuthenticationProvider -v="Azure Active Directory" -s $Address --apikey=$system
```

Note that `AzureADAuthority` does not contain `https://`. 👍

There is one minor issue in that I am not able to delete the administrative API key after this because of the `AuthenticationProvider` mode.  I have to delete it from the dashboard upon login.  All things considered this is a minor nit as everything else works perfectly now, but if there's any thought you could lend towards allowing this to be deleted via automation it would be appreciated.

Thank you again for all your help!  This is truly wonderful. 🙏

**Reply by mloskot** [reply]
First, thanks for the detailed discussion which seems to be reproducible for any non-interactive Seq deployment with Entra ID authentication.

Am I correct seeing that these `AzureAD*` settings are not documented anywhere at https://datalust.co/docs/ ?

**Reply by Mike-E-angelo** [reply]
Ah indeed, @mloskot, to add to the churn and fun, `AzureAD*` has since been replaced with `EntraID*`.  From my current script:

```
        seqcli setting set -n=EntraIDAuthority -v="login.microsoftonline.com" -s $Address --apikey=$system
        seqcli setting set -n=EntraIDClientId -v="$application" -s $Address --apikey=$system
        seqcli setting set -n=EntraIDClientKey -v="$secret" -s $Address --apikey=$system
        seqcli setting set -n=EntraIDTenantId -v="$TenantId" -s $Address --apikey=$system
        seqcli setting set -n=AuthenticationProvider -v="Microsoft Entra ID" -s $Address --apikey=$system
```

**Reply by nblumhardt** [reply]
Thanks for the nudge RE the documentation gap! Until we fix that, although coverage isn't quite 100%, you can find the bulk of these settings described in:

https://github.com/datalust/seq-api/blob/dev/src/Seq.Api/Model/Settings/SettingName.cs#L27

**Reply by mloskot** [reply]
@nblumhardt That will help too, thank you!
